### ISSUE TITLE
Circular dependencies

### ISSUE THREAD

**Issue by samfux84** [spack/spack#3664]
Hi,

I would like to build python with spack, but it seems that it has circular dependencies:

python -> tk -> libx11 -> libxcb -> python

How can this be resolved ?

Best regards

**Comment by tgamblin** [comment]
Can you try this with #2548 merged?

**Comment by samfux84** [comment]
Do I need to do a fresh spack installation to merge the pull request #2548 ?

[sfux@develop02 spack]$ git fetch origin pull/2548/head:master
fatal: Refusing to fetch into current branch refs/heads/master of non-bare repository
[sfux@develop02 spack]$

**Comment by tgamblin** [comment]
@samfux84: you shouldn't.  I think you're getting that error because you're trying to fetch directly into a checked out branch.

Try:
```
# fetch into a new branch
git fetch origin pull/2548/head:pull-2548

# check out a new branch based on your current one so you can revert later
git co -b test-2548 

# merge the PR branch into this new branch.
git merge pull-2548
```

**Comment by samfux84** [comment]
@tgamblin: Thank you for helping me. I am still not that used to git yet. The commands that you provided worked. I could merge the pull request #2548 with my current branch.

I am afraid that I still did not fully understand how this resolves the circular dependencies.

I usually build all the dependencies first, and then the main package, such that there is only 1 installation done at a time (no automatic dependency resolving). Where should I start with the chain

python -> tk -> libx11 -> libxcb -> python

Any help is appreciated.

**Comment by tgamblin** [comment]
I would first try `spack install python+tk` on that branch. 

The issue arises from our concretizer being a bit naive.  It can't resolve the fact that it first needs to build `python ~tk` to build a dependency of `python+tk`.  That branch allows us to deal with build dependencies separately from run dependencies so that we can at least start to deduce chains like this.

**Comment by samfux84** [comment]
Thank you for your reply.

I tried the command you suggested, but it fails:
```

[sfux@develop02 tools]$ spack install python@2.7.13+tk@8.6.6
==> Installing python
==> bzip2 is already installed in /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/bzip2-1.0.6-ch5qiinczzmltn4bsi354xreyyoavtaj
==> ncurses is already installed in /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/ncurses-6.0-wvgtxi53n6canuianx6bgjkyq3l3flt6
==> zlib is already installed in /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/zlib-1.2.11-rkzkdjjffuzl55dql2pm2o6dmiq2shu3
==> openssl is externally installed in /usr
==> sqlite is already installed in /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/sqlite-3.8.5-qhtgyb7nwxyqq4ti2ea6vx52b7nlt237
==> tcl is already installed in /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/tcl-8.6.6-xsyci3gk3n6wzfrnjm5m2ijez5wfyyra
==> Installing tk
==> tcl is already installed in /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/tcl-8.6.6-xsyci3gk3n6wzfrnjm5m2ijez5wfyyra
==> Fetching http://prdownloads.sourceforge.net/tcl/tk8.6.6-src.tar.gz
######################################################################## 100.0%
==> Staging archive: /local/cluster/spack/var/spack/stage/tk-8.6.6-uikto6jjrwhb4xykmxujtwzhyph4zlvd/tk8.6.6-src.tar.gz
==> Created stage in /local/cluster/spack/var/spack/stage/tk-8.6.6-uikto6jjrwhb4xykmxujtwzhyph4zlvd
==> Ran patch() for tk
==> Building tk [AutotoolsPackage]
==> Executing phase : 'autoreconf'
==> Executing phase : 'configure'
==> Executing phase : 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j8'
/local/cluster/spack/lib/spack/spack/build_systems/autotools.py:275, in build:
     270      def build(self, spec, prefix):
     271          """Makes the build targets specified by
     272          :py:attr:``~.AutotoolsPackage.build_targets``
     273          """
     274          with working_dir(self.build_directory):
  >> 275              inspect.getmodule(self).make(*self.build_targets)

See build log for details:
  /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/spack-build.out
```
The build can not find the X11 header files:

```
/local/cluster/spack/lib/spack/env/gcc/gcc -c -O2  -pipe  -Wall -fPIC -DBUILD_tk -I/local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../unix -I/local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic -I/local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../bitmaps -I/local/scratch/sfux/spack-stage/spack-stage-JCygqL/tcl8.6.6/generic -I/local/scratch/sfux/spack-stage/spack-stage-JCygqL/tcl8.6.6/unix  -DPACKAGE_NAME=\"tk\" -DPACKAGE_TARNAME=\"tk\" -DPACKAGE_VERSION=\"8.6\" -DPACKAGE_STRING=\"tk\ 8.6\" -DPACKAGE_BUGREPORT=\"\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DUSE_THREAD_ALLOC=1 -D_REENTRANT=1 -D_THREAD_SAFE=1 -DHAVE_PTHREAD_ATTR_SETSTACKSIZE=1 -DHAVE_PTHREAD_ATFORK=1 -DTCL_THREADS=1 -DMODULE_SCOPE=extern\ __attribute__\(\(__visibility__\(\"hidden\"\)\)\) -DHAVE_HIDDEN=1 -DHAVE_CAST_TO_UNION=1 -DTCL_SHLIB_EXT=\".so\" -DNDEBUG=1 -DTCL_CFG_OPTIMIZED=1 -D_LARGEFILE64_SOURCE=1 -DTCL_WIDE_INT_IS_LONG=1 -DHAVE_SYS_TIME_H=1 -DTIME_WITH_SYS_TIME=1 -DHAVE_INTPTR_T=1 -DHAVE_UINTPTR_T=1 -DHAVE_PW_GECOS=1 -DXKEYCODETOKEYSYM_IS_DEPRECATED=1      -DTCL_NO_DEPRECATED  -DUSE_TCL_STUBS /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkBusy.c
In file included from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/ttk/ttkStubLib.c:9:0:
/local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tk.h:96:25: fatal error: X11/Xlib.h: No such file or directory
 #   include <X11/Xlib.h>
                         ^
compilation terminated.
In file included from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkPort.h:21:0,
                 from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkInt.h:19,
                 from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkArgv.c:14:
/local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tk.h:96:25: fatal error: X11/Xlib.h: No such file or directory
 #   include <X11/Xlib.h>
                         ^
compilation terminated.
In file included from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkPort.h:21:0,
                 from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkInt.h:19,
                 from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkStubLib.c:14:
/local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tk.h:96:25: fatal error: X11/Xlib.h: No such file or directory
 #   include <X11/Xlib.h>
                         ^
compilation terminated.
In file included from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkPort.h:21:0,
                 from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkInt.h:19,
                 from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkBitmap.c:15:
/local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tk.h:96:25: fatal error: X11/Xlib.h: No such file or directory
 #   include <X11/Xlib.h>
                         ^
compilation terminated.
In file included from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkPort.h:21:0,
                 from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkInt.h:19,
                 from /local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tkBusy.c:15:
/local/scratch/sfux/spack-stage/spack-stage-tiFl3P/tk8.6.6/unix/../generic/tk.h:96:25: fatal error: X11/Xlib.h: No such file or directory
 #   include <X11/Xlib.h>
                         ^
compilation terminated.
```

**Comment by adamjstewart** [comment]
This should have been fixed in another branch. I modified `tk` to always depend on X11 unconditionally. Can you run `spack spec python+tk` to see whether it is building `tk+X` or `tk~X`?

**Comment by samfux84** [comment]
```
[sfux@develop02 tools]$ spack spec python@2.7.13+tk@8.6.6 
Input spec
--------------------------------
python@2.7.13,8.6.6+tk

Normalized
--------------------------------
python@2.7.13,8.6.6+tk
    ^bzip2
    ^ncurses
    ^openssl
        ^zlib
    ^readline
    ^sqlite
    ^tcl
    ^tk

Concretized
--------------------------------
python@2.7.13%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" +tk~ucs4 arch=linux-centos7-x86_64 
    ^bzip2@1.0.6%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" +shared arch=linux-centos7-x86_64 
    ^ncurses@6.0%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" ~symlinks arch=linux-centos7-x86_64 
    ^openssl@1.0.1e%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx"  arch=linux-centos7-x86_64 
    ^readline@6.3%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx"  arch=linux-centos7-x86_64 
    ^sqlite@3.8.5%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx"  arch=linux-centos7-x86_64 
    ^tcl@8.6.6%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx"  arch=linux-centos7-x86_64 
        ^zlib@1.2.11%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" +pic+shared arch=linux-centos7-x86_64 
    ^tk@8.6.6%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" ~X arch=linux-centos7-x86_64 
```

**Comment by samfux84** [comment]
```
[sfux@develop02 tools]$ spack spec python+tk
Input spec
--------------------------------
python+tk

Normalized
--------------------------------
python+tk
    ^bzip2
    ^ncurses
    ^openssl
        ^zlib
    ^readline
    ^sqlite
    ^tcl
    ^tk

Concretized
--------------------------------
python@2.7.13%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" +tk~ucs4 arch=linux-centos7-x86_64 
    ^bzip2@1.0.6%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" +shared arch=linux-centos7-x86_64 
    ^ncurses@6.0%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" ~symlinks arch=linux-centos7-x86_64 
    ^openssl@1.0.1e%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx"  arch=linux-centos7-x86_64 
    ^readline@6.3%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx"  arch=linux-centos7-x86_64 
    ^sqlite@3.8.5%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx"  arch=linux-centos7-x86_64 
    ^tcl@8.6.6%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx"  arch=linux-centos7-x86_64 
        ^zlib@1.2.11%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" +pic+shared arch=linux-centos7-x86_64 
    ^tk@8.6.6%gcc@4.8.5 cflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cppflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" cxxflags="-O3 -ftree-vectorize -mavx -fPIC -march=corei7-avx" ~X arch=linux-centos7-x86_64 

[sfux@develop02 tools]$ 
```

**Comment by luigi-calori** [comment]
I have  the same problem of @samfux84 some  week ago I tried to merge #2548 and it merged, and I was able to build python+tk ^tk+X  ... unfortunately, I found out that merging #2548 had a side effect of not being able to build other python packages

Now I tried the merge but now it seems not mergeable and github report [conflicts](https://github.com/LLNL/spack/pull/2548/conflicts)
Am I missing some step in right approach to merging? 
One conflict is related to this [commit](https://github.com/LLNL/spack/commit/1d1a14dbe9cfc512cedff7430b694df87b2cd5ee#diff-c8f14b8e4ac7661fb9ca8ecc00bc2ebc)

Any hints on a workaround?  Is there a way to help in having this problem solved?

Thanks in advance

**Comment by adamjstewart** [comment]
@samfux84 You must be using an older version of Spack, because `tk` now unconditionally depends on `libx11`.

**Comment by luigi-calori** [comment]
Uh... that could also explain why  I' m not able to merge  2548 and why also github reports conflicts

**Comment by adamjstewart** [comment]
#2548 also has conflicts. Every time another PR affects the same lines of the same file, it creates a conflict. #2548 will need to be updated to resolve these conflicts. You can do this by hand with a merge or rebase.

**Comment by luigi-calori** [comment]
MM.. apart the fact that I' m not sure to be able to correctly edit the conflicts, how could then keep the patch? Is there a way to submit a PR on PR?

**Comment by adamjstewart** [comment]
Only if the PR contributor allows you to push to the same branch. It's probably best to wait for @scheibelp to update it. But I agree that building `python+tk` is important. Better concretization is the primary goal for the next Spack release, so I'm confident that it will be merged soon.

**Comment by scheibelp** [comment]
> #2548 also has conflicts

I can fix the conflicts (with develop) today. These conflicts appeared over the last 2 days. 

> I found out that merging #2548 had a side effect of not being able to build other python packages

@luigi-calori  Would you be able to mention which packages fail for you in the PR (once I get it consistent with develop)?

**Comment by luigi-calori** [comment]
Thanks @scheibelp 
If I remember well one was py-cffi
As  soon as the conflicts are solved, I can try to merge and use it to build my stack ( paramiko was one of the module that I need along with python+tk )

**Comment by scheibelp** [comment]
I have updated #2548 to resolve merge conflicts with develop.

However, the py-cffi example showed me that python modules which depended on py-setuptools to build are broken for that PR. I have a more-detailed post on my thoughts in the PR conversation.

I added a commit which resolves the issue for py-cffi specifically, but I will likely end up reverting that and pursuing a different fix.

**Comment by samfux84** [comment]
@adamjstewart: For my current test installation (~ 80 packages), I am using the master branch + merged PR's: 2548, 2972, 3388, 3543

Would you recommend to use the development branch ?

I went for the master branch, because I am planing to use spack for managing almost the entire software stack of our HPC cluster (I thought first that the development branch would change often, and therefore it would be better to go with the master branch).

**Comment by adamjstewart** [comment]
@samfux84 We use Spack to manage our entire software stack at [ANL's LCRC](http://lcrc.anl.gov/), serving around 300 active users with diverse software needs. We've been keeping up-to-date with Spack's develop branch for the last 1.5 years with only minor hiccups. Yes, things break often, but not as often as things get fixed. Spack is still at the point where I would recommend most users stick with develop over master. Maybe in a year or two things will settle down to the point where master is fine. Once we get better software testing, we plan on having more frequent releases.

If you still want to stick with the master branch and keep merging in fixes, you can add #3517 which fixes your problem, and also #3544 which fixes a bug that #3517 introduced.

**Comment by samfux84** [comment]
@adamjstewart Thank you for your reply. It would then actually make sense for me to also switch to the development branch. I will try to rebuild the ~80 packages from my test installation and then try to build python+tk and then provide feedback.

**Comment by samfux84** [comment]
Switched to development branch + PR's: 2548, 2972, 3735. After reinstalling all packages I tried again to install Python and this time it worked, as tcl and tk are not required to build Python (are they only used if +tk is specified ?):

```
[sfux@develop02 tools]$ spack install python@2.7.13 ^bzip2@1.0.6 ^ncurses@6.0 ^openssl@1.0.1e ^readline@7.0 ^sqlite@3.8.5 ^zlib@1.2.11
==> Installing python
==> bzip2 is already installed in /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/bzip2-1.0.6-ainh3dktbckz5ke2jckhuo75ydapuf3n
==> ncurses is already installed in /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/ncurses-6.0-yqf6gqrtuhwtfs6rlsfvpmpvcu63cgo4
==> zlib is already installed in /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/zlib-1.2.11-vnn7gj6byrvbs6qjx3nc57gnynhsv2qk
==> openssl is externally installed in /usr
==> sqlite is already installed in /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/sqlite-3.8.5-ali6my462us54untd462rzrdpdvbl2kv
==> readline is already installed in /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/readline-7.0-p5j542rky6ftmlge27u2fk3m5hsbqn2w
==> Fetching http://www.python.org/ftp/python/2.7.13/Python-2.7.13.tgz
######################################################################## 100.0%
==> Staging archive: /local/cluster/spack/var/spack/stage/python-2.7.13-jewe3ea2reglw6vt3cw7sbxhp7d3nrqj/Python-2.7.13.tgz
==> Created stage in /local/cluster/spack/var/spack/stage/python-2.7.13-jewe3ea2reglw6vt3cw7sbxhp7d3nrqj
==> Applied patch ncurses.patch
==> Ran patch() for python
==> Building python [Package]
==> Executing phase : 'install'
==> Successfully installed python
  Fetch: 1.06s.  Build: 1m 44.95s.  Total: 1m 46.01s.
[+] /local/cluster/spack/apps/linux-centos7-x86_64/gcc-4.8.5/python-2.7.13-jewe3ea2reglw6vt3cw7sbxhp7d3nrqj
[sfux@develop02 tools]$
```

**Comment by adamjstewart** [comment]
> tcl and tk are not required to build Python (are they only used if +tk is specified ?)

Yes. You can see this in the `depends_on` directives.

**Comment by scheibelp** [comment]
If you don't need to build ```python+tk``` and just want to build ```python@version```, then you should not need to include #2548.

**Comment by alalazo** [comment]
Following the discussion this seems solved to me.
